### PR TITLE
Selective cache clear

### PR DESCRIPTION
**Pull request by o-**
based on #502 , only clear cache idx which belong to a particular mkenv.

### CODE REVIEW

File: rir/src/ir/BC_inc.h [hunk 1]
```diff
@@ -121,6 +121,10 @@ class BC {
         PoolIdx poolIndex;
         CacheIdx cacheIndex;
     };
+    struct CacheIdxs {
```
Comment (charig) [inline rir/src/ir/BC_inc.h]: CacheRange?

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 2]
```diff
@@ -451,26 +451,88 @@ class CachePosition {
...
+                        break;
+                    uniqueNumbers.emplace(key.first, uniqueNumbers.size());
+                }
+            }
+        }
+        globalEnvsCacheSize_ = uniqueNumbers.size();
+
+        // Heuristic: after how many loads does it pay off to cache a variable?
+        // It depends on the size of the environment, since the smaller the
+        // environment, the faster the lookup.
```
Comment (charig) [inline rir/src/compiler/translations/pir_2_rir/allocators.h]: Why the smaller the faster? It is a hash table...
Oth, the code defines bigger limits for smaller environment
And then, why for non-global environments we cache even if there is only one load while for global we do not?
Also how the heuristic treat loop lds is unclear...

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 3]
```diff
@@ -451,26 +451,88 @@ class CachePosition {
...
+                return 99999; // This env seems empty, caching is just a waste.
+            else if (s == 1)
+                return 3;
+            else if (s < 5)
+                return 2;
+            return 1;
+        };
+        for (const auto& env : found) {
+            if (!Env::Cast(env.first)) {
+                envCachePositions[env.first].first = uniqueNumbers.size();
```
Comment (charig) [inline rir/src/compiler/translations/pir_2_rir/allocators.h]: envCacheRanges?

File: rir/src/compiler/translations/pir_2_rir/allocators.h [hunk 4]
```diff
@@ -451,26 +451,88 @@ class CachePosition {
...
+            }
+        });
+
+        // First all global envs
+        for (const auto& env : found) {
+            auto e = Env::Cast(env.first);
+            if (e && e->rho != R_BaseEnv && e->rho != R_BaseNamespace) {
+                for (const auto& key : env.second) {
+                    if (key.second <= 2)
+                        continue;
```
Comment (charig) [inline rir/src/compiler/translations/pir_2_rir/allocators.h:480]: What if it is inside a loop?

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 2 above)
Comment (o-) [reply rir/src/compiler/translations/pir_2_rir/allocators.h]: > Why the smaller the faster? It is a hash table...

no, its a list. lookup is a linear search

> Oth, the code defines bigger limits for smaller environment

yes, the smaller the env, the more lookups are needed to win anything by caching

> And then, why for non-global environments we cache even if there is only one load while for 
> global we do not?

well if you think about it, for local envs the number of bindings we statically see is roughly equal to the size of the environment. but for a global env, that is certainly not true, as the global env will have many more unrelated bindings. so there this heuristic does not work.

> Also how the heuristic treat loop lds is unclear...

well it doesn't, but in practice it still seems to work quite nicely

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 4 above)
Comment (o-) [reply rir/src/compiler/translations/pir_2_rir/allocators.h:480]: i know, still a somewhat useful proxy.

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 2 above)
Comment (charig) [reply rir/src/compiler/translations/pir_2_rir/allocators.h]: I didn't get it. You were talking about the standard lookup. I understood it as the binding cache lookup.
I would add to the comment ...the smaller the env, the faster the (standard|noncached) lookup. 
Oth, this is not taking into account that the lookup could need to go through a hierarchy of envs.

**Review by charig — APPROVED** [review]

File: rir/src/compiler/translations/pir_2_rir/allocators.h (hunk 2 above)
Comment (o-) [reply rir/src/compiler/translations/pir_2_rir/allocators.h]: > I would add to the comment ...the smaller the env, the faster the (standard|noncached) lookup.

ok, I'll make it more clear

> Oth, this is not taking into account that the lookup could need to go through a hierarchy of envs.

sure, but then we don't cache anyway.

**Pull request merged**